sdks/python/apache_beam/runners/interactive/interactive_environment.py

```diff
@@ -221,3 +224,20 @@ def is_terminated(self, pipeline):
     if result:
       return runner.PipelineState.is_terminal(result.state)
     return True
+
+  def track_user_pipelines(self):
+    """Tracks all user-defined pipeline instances.
```

Comment:
Please explain what it means by "tracking" user pipelines.

Reply:
Rephrased to
```
    """Record references to all user-defined pipeline instances watched in
    current environment.

    Current static global singleton interactive environment holds references to
    a set of pipeline instances defined by the user in the watched scope.
    Interactive Beam features could use the references to determine if a given
    pipeline is defined by user or implicitly created by Beam SDK or runners,
    then handle them differently.

    This is invoked every time a PTransform is to be applied if the current
    code execution is under ipython due to the possibility that any user-defined
    pipeline can be re-evaluated through notebook cell re-execution at any time.

    Each time this is invoked, the tracked user pipelines are refreshed to
    remove any pipeline instances that are no longer in watched scope. For
    example, after a notebook cell re-execution re-evaluating a pipeline
    creation, the last pipeline reference created by last evaluation will not be
    in watched scope anymore.
    """
```